### PR TITLE
Docathon! Docs for discrete pulse functions 

### LINKED ISSUE #3392: Add mathematical definition of pulses in pulse_lib documentation
### What is the expected enhancement?
Mathematical definitions of pulses should be added in the `pulse_lib` documentation.

### PR DESCRIPTION
**Pull request by DanPuzzuoli**
Closes #3392

### Summary

Adding math expressions for the functions being sampled in `discrete.py`. 

### Details and comments

1. Updated doc strings in `discrete.py` to include math expressions, along with a couple of jupyter-execute sections for plotting pulses with a simple shape but whose mathematical expression may not be immediately recognizable (or at least weren't to me).
2. Code provided by @SooluThomas: Updated `pulse.rst` and `pulse/__init__.py` to include a section in the sphinx pulse toc called Pulse Library, which links to the `discrete` module.

### CODE REVIEW

File: qiskit/pulse/pulse_lib/discrete.py [hunk 1]
```diff
@@ -102,16 +142,35 @@ def sawtooth(duration: int, amp: complex, period: float = None,
...
+    For :math:`A=` ``amp``, :math:`T=` ``period``, and :math:`\phi=` ``phase``,
+    applies the `midpoint` sampling strategy to generate a discrete pulse sampled from
+    the continuous function:
+
+    .. math::
 
-    Applies `midpoint` sampling strategy to generate discrete pulse from continuous function.
+        f(x) = A \left(-2\left|\text{sawtooth}(x, A, T, \phi)\right| + 1\right)
+
+    This corresponds to a sine wave with linear ramps.
```
Comment (taalexander) [inline qiskit/pulse/pulse_lib/discrete.py]: I don't think this is a "sine wave", if anything it is something like

```suggestion
    This a non-sinusoidal wave with linear ramping.
```

File: qiskit/pulse/pulse_lib/discrete.py [hunk 2]
```diff
@@ -195,13 +282,20 @@ def gaussian(duration: int, amp: complex, sigma: float, name: Optional[str] = No
...
+    .. math::
+
+        f(x) = A\frac{2(x - \mu)}{\sigma^2}\exp\left(\left(\frac{x - \mu}{\sigma}\right)^2 \right)
+
+    i.e. the derivative of the Gaussian function, with center :math:`\mu=` ``duration/2``.
 
     Args:
         duration: Duration of pulse. Must be greater than zero.
-        amp: Pulse amplitude at `center`.
+        amp: Pulse amplitude of corresponding Gaussian at center ``duration/2``.
```
Comment (taalexander) [inline qiskit/pulse/pulse_lib/discrete.py]:
```suggestion
        amp: Pulse amplitude of corresponding Gaussian at the pulse center (``duration/2``).
```

File: qiskit/pulse/pulse_lib/discrete.py [hunk 3]
```diff
@@ -299,23 +430,34 @@ def gaussian_square(duration: int, amp: complex, sigma: float,
...
         sigma: Width (standard deviation) of pulse.
-        beta: Y correction amplitude. For the SNO this is $\beta=-\frac{\lambda_1^2}{4\Delta_2}$.
-            Where $\lambds_1$ is the relative coupling strength between the first excited and second
-            excited states and $\Delta_2$ is the detuning between the respective excited states.
+        beta: Y correction amplitude. For the SNO this is
+              :math:`\beta=-\frac{\lambda_1^2}{4\Delta_2}`. Where :math:`\lambda_1` is the
+              relative coupling strength between the first excited and second excited states
+              and :math:`\Delta_2` is the detuning between the respective excited states.
         name: Name of pulse.
         zero_ends: If True, make the first and last sample zero, but rescale to preserve amp.
```
Comment (SooluThomas) [inline qiskit/pulse/pulse_lib/discrete.py]:
```suggestion
        zero_ends: If ``True``, make the first and last sample zero, but rescale to preserve amp.
```

**Comment by SooluThomas** [comment]
Thank you @DanPuzzuoli

File: qiskit/pulse/__init__.py [hunk 4]
```diff
@@ -74,6 +74,14 @@
    LoConfig
    LoRange
 
+Pulse Library
+=============
+
+.. autosummary::
+   :toctree: ../stubs/
+
+   qiskit.pulse.pulse_lib.discrete
```
Comment (lcapelluto) [inline qiskit/pulse/__init__.py]: tilde works here too!
```suggestion
   ~qiskit.pulse.pulse_lib.discrete
```

File: qiskit/pulse/pulse_lib/discrete.py [hunk 5]
```diff
@@ -166,18 +239,32 @@ def sin(duration: int, amp: complex, freq: float = None,
...
+
+    For :math:`A=` ``amp`` and :math:`\sigma=` ``sigma``, applies the `midpoint` sampling strategy
+    to generate a discrete pulse sampled from the continuous function:
 
-    Centered at `duration/2` and zeroed at `t=0` and `t=duration` to prevent large
-    initial/final discontinuities.
+    .. math::
 
-    Applies `midpoint` sampling strategy to generate discrete pulse from continuous function.
+        f(x) = A\exp\left(\left(\frac{x - \mu}{\sigma}\right)^2 \right),
```
Comment (lcapelluto) [inline qiskit/pulse/pulse_lib/discrete.py]: I think this should be 
```suggestion
        f(x) = A\exp\left(\left(\frac{x - \mu}{2\sigma}\right)^2 \right),
```

File: qiskit/pulse/pulse_lib/discrete.py [hunk 6]
```diff
@@ -195,13 +282,20 @@ def gaussian(duration: int, amp: complex, sigma: float, name: Optional[str] = No
...
-    r"""Generates unnormalized gaussian derivative `SamplePulse`.
+    r"""Generates unnormalized gaussian derivative :class:`~qiskit.pulse.SamplePulse`.
+
+    For :math:`A=` ``amp`` and :math:`\sigma=` ``sigma`` applies the `midpoint` sampling strategy
+    to generate a discrete pulse sampled from the continuous function:
 
-    Applies `midpoint` sampling strategy to generate discrete pulse from continuous function.
+    .. math::
+
+        f(x) = A\frac{2(x - \mu)}{\sigma^2}\exp\left(\left(\frac{x - \mu}{\sigma}\right)^2 \right)
```
Comment (lcapelluto) [inline qiskit/pulse/pulse_lib/discrete.py]: If the gaussian func has x**2/(2sigma), then this is
```suggestion
        f(x) = A\frac{(x - \mu)}{\sigma^2}\exp\left(\left(\frac{x - \mu}{\sigma}\right)^2 \right)
```

**Comment by DanPuzzuoli** [comment]
Nice catch @lcapelluto, just committed

**Pull request merged**